### DISCUSSION TITLE
How to get current bundle configuration in `prependExtensionConfig`?

### DISCUSSION THREAD

**Discussion by mantassimkunas** [symfony/symfony#57951] [Q&A]
In my custom bundle, I want to configure another bundle but to do so I need a few values that are provided by the application via my bundle configuration.

I can configure other bundles with `preprendExtensionConfig` but at this stage I do not yet have the processed configuration of my bundle. Is there a better approach?

**Comment by stof** [comment]
Doing so it quite unreliable, because even if you get it (using `$container->getExtensionConfig()`), you have no guarantee at all that this configuration is the full one yet (maybe another bundle also wants to prepend some configuration for your bundle, and it will be called after yours).

**Comment by yceruto** [comment]
There is only one chance to get the processed config and still prepend configs for other bundles without facing the issue explained by @stof above.

In your bundle's load extension method, you can prepend config for other bundles, but either the given `ContainerConfigurator $container` or `ContainerBuilder $builder` are based on a temporary container and the container merging mechanism won't prepend that config properly but will append it, overriding your app config. So, never use them to prepend config at this stage.

What you can do is store the container builder that the `prependExtension` method provides you and reuse it (this is the main one).

Take this as an example:
```php
class MyOwnBundle extends AbstractBundle
{
    private ContainerBuilder $builder;

    public function prependExtension(ContainerConfigurator $container, ContainerBuilder $builder): void
    {
        $this->builder = $builder;
    }

    public function loadExtension(array $config, ContainerConfigurator $container, ContainerBuilder $builder): void
    {
        $this->builder->prependExtensionConfig('third_party', [
            // prepend config based on the processed $config ...
        ]);
    }
}
```
The important thing here is that you have to ensure that `MyOwnBundle` is registered before `third_party` bundle/extension in `bundles.php`. Otherwise, it will be too late for `third_party` to know about your custom prepended config. That's the tricky part... but maybe you could enforce that by checking the order during the prepend extension phase and throw an exception?

Now, there’s always the possibility that other people start using the same trick as you for the same bundle, and then you’re all in the same situation as before 😅 so use this quietly.

**Reply by mantassimkunas** [reply]
Thank you for such a detailed reply.

It's becoming apparent to me now that this is more complex than I initially hoped. There's a lot of hidden knowledge here, and to be perfectly honest, the overall approach seems a bit... hacky.

I was wondering - why not just ship a recipe with my bundle which places a `my_own_bundle.yaml` file in the app's `config/packages` and configures other bundles? I want to configure the cache, messenger, twig and other components from within my bundle and it looks like storing that configuration as part of the application-owned yaml config is the simplest approach. After all, the recipes are versioned and I can always ship an updated version, and the application has the ability to override those settings at any time (same as with the `prependExtensionConfig` approach).

**Reply by yceruto** [reply]
If it's all about preconfig, you don't need to worry about the app config because the app config will always take precedence over the preconfigured one. This, of course, depends on the config definition itself and whether it allows prototype nodes with a merging strategy. But in principle, if they are single configs, you can safely prepend one in your bundle, and if there's a concrete config at the app level, it will take priority.

**Reply by msimkunas** [reply]
My goal was to have a self contained bundle that preconfigures everything itself. But with recipes it’s still possible to ship config file updates so the fact that the application owns this file is probably not much of an issue. There’s actually an upside because I *need* some values that are only available on the application side.

**Reply by msimkunas** [reply]
To be clear, I was referring to, for example, configuring new cache pools under the framework.cache config key in my_own_bundle.yaml.

**Reply by stof** [reply]
The simplest solution is to define a new cache pool in your bundle without relying on asking FrameworkBundle to do it. This is exactly what I'm doing in https://github.com/Incenteev/hashed-asset-bundle/blob/b4094e85149b558eee81a2ac8c2e0f33c3bfe92b/src/Resources/config/cache.xml#L35-L37

FrameworkBundle already supports such extension point by using a compiler to configure the parts relying on knowing the list of available pools (allowing other bundles to also define some services tagged as `cache.pool`)

**Reply by msimkunas** [reply]
@stof Hmm, this works because you can just tag that service with the appropriate tag. However, what about if you need to configure something like Twig namespaces? Shipping a bundle config file which does all the necessary setup under the twig config key is probably the easiest approach then.

**Comment by yceruto** [comment]
Another alternative is to use a compiler pass that changes or creates services based on some config parameters that you previously set in your bundle extension stage. It's way more complex but still a possibility.

**Reply by msimkunas** [reply]
I thought about this but yes, looks like it’s too complex. So far, shipping my own yaml recipe which preconfigures other bundles seems easiest. The only downside is that this exposes the preconfig to the application but like you already mentioned in another comment, app configuration always takes precedence, so there’s no real encapsulation of config anyway. With a shippable config/packages/my_own_bundle.yaml I don’t have to worry about access to existing parameters, env variables, etc because I can just provide sane defaults and the application can override it as necessary.
